### DISCUSSION TITLE
[Native DLL] Opaque structs

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3027] [General]
Should I map it to `IntPtr` or an empty record?

**Comment by SunSerega** [comment]
That's for you to decide - anything will work.

There is only one thing I can suggest:
You have probably already seen HWND (Handle-of-WiNDow) parameters to WinAPI functions. They are too opaque structs but are usually just described using an `IntPtr`.
But there could be many other unrelated and incompatible handles, from other APIs. The API itself wouldn't stop you from passing the handle of the wrong type. But you can make your own record.
Just ensure it has the same size as `IntPtr` - that's all that matters. For that you can put a single private field of type `IntPtr` (or maybe `pointer`, etc.) inside.
And then, you know, since it's a type specifically for window handles, how about adding an OOP interface to it, containing some of the most common WinAPI functions?

Here is how I've done it when I needed something like that:
https://github.com/SunSerega/AutoMaximize/blob/522bd704ed598e1805b9e5cef05b55e76867eec8/WinAPI.pas#L36
And then implement those methods by calling external methods and passing in self:
https://github.com/SunSerega/AutoMaximize/blob/522bd704ed598e1805b9e5cef05b55e76867eec8/WinAPI.pas#L106

Or, another example: In OpenGL, buffers are referenced by not a pointer-sized, but 32-bit descriptor (aka. handle/name/id/etc.): https://github.com/SunSerega/POCGL/blob/e1b884ae85c97613c8e095c0f714b63666d9f2e1/Packing/Template/LowLvl/OpenGL/Types.Interface.template#L26548
Here I didn't add any methods because they have a very complex initialization procedure (you need to query the driver to get function pointers) and many of them take in multiple handles of multiple types.
